java/arcs/core/storage/StorageProxy.kt

```diff
@@ -37,57 +37,50 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperation, T>(
     initialCrdt: CrdtModel<Data, Op, T>
 ) {
     private val log = TaggedLog { "StorageProxy" }
-    private val mutex = reentrantLock()
+    private val handlesMutex = Mutex()
     private val handles: MutableSet<Handle<Data, Op, T>> = mutableSetOf()
+    private var hasReaders = false
+    private val syncMutex = Mutex()
     private val crdt = initialCrdt
     private val waitingSyncs: MutableList<CompletableDeferred<ValueAndVersion<T>>> = mutableListOf()
     private var synchronized = false
-    private var listenerAttached = false
-    private var hasReaders = false
 
     private val store = storeEndpointProvider.getStorageEndpoint()
 
+    init {
+        store.setCallback(ProxyCallback(::onMessage))
+    }
+
     /**
      * Connects a handle. If the handle is readable, it will receive the configured callbacks.
      */
     suspend fun registerHandle(handle: Handle<Data, Op, T>): VersionMap {
         // non-readers don't get callbacks, return early
-        if (!handle.canRead) return mutex.withLock { crdt.versionMap.copy() }
+        if (!handle.canRead) return run { crdt.versionMap.copy() }
 
         log.debug { "Registering handle: $handle" }
 
-        mutex.withLock {
-            if (!listenerAttached) {
-                log.debug { "Attaching listener to store: $store" }
-                store.setCallback(ProxyCallback(::onMessage))
-                listenerAttached = true
-            }
-
+        val firstReader = handlesMutex.withLock {
             handles.add(handle)
+            val result = !hasReaders
+            hasReaders = true
+            result
+        }
 
-            // Change to synchronized mode as soon as we get any read handles and send a request to get
-            // the full model (once).
-            if (!this.hasReaders) {
-                this.requestSynchronization()
-                this.hasReaders = true
-            }
+        val (hasSynced, versionMap) = syncMutex.withLock { synchronized to crdt.versionMap.copy() }
 
-            // If a handle registers after we've received the full model, notify it immediately.
-            if (this.synchronized) {
-                coroutineScope {
-                    launch { handle.callback?.onSync() }
-                }
-            }
-            return crdt.versionMap.copy()
-        }
+        if (firstReader) requestSynchronization()
+        else if (hasSynced) handle.callback?.onSync()
```

Comment:
This should be launched, user's code should not have any possibility to block storage proxy's operations.